### PR TITLE
fix: Unauthenticated api throttling

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

We were throttling certain requests such as email signup / password reset based on the request which is a bit loose given we almost always have something we can identify the intended user with (email or uuid).

Also we didn't set a scope which means that one throttler could be affecting the other. We have really harsh rate limits on password resets which then in turn affect anything else scoped to that user.

## Changes

* Adds a custom throttler that tries to find the authed user or email|uuid in the payload, hashes it and falls back to the ident.

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

File: posthog/api/authentication.py [hunk 1]
```diff
@@ -190,6 +190,7 @@ class LoginViewSet(NonCreatingViewSetMixin, viewsets.GenericViewSet):
     queryset = User.objects.none()
     serializer_class = LoginSerializer
     permission_classes = (permissions.AllowAny,)
+    # NOTE: Throttling is handled by the `axes` package
```
Comment (benjackwhite) [inline posthog/api/authentication.py:193]: Added a note here as I wasn't sure why we had no throttler for logging in

File: posthog/api/test/test_authentication.py [hunk 2]
```diff
@@ -434,6 +434,23 @@ def test_cant_reset_more_than_six_times(self):
         # Three emails should be sent, fourth should not
         self.assertEqual(len(mail.outbox), 6)
 
+    def test_is_rate_limited_on_email_not_ip(self):
```
Comment (pauldambra) [inline posthog/api/test/test_authentication.py:437]: i'm lazily assuming there's a test in here that shows the reverse (if we have the same IP but varying addresses then we don't limit (or we're slower to limit))

File: posthog/rate_limit.py [hunk 3]
```diff
@@ -222,6 +223,26 @@ def get_bucket_key(self, request):
         return ident
 
 
+class UserOrEmailRateThrottle(SimpleRateThrottle):
```
Comment (pauldambra) [inline posthog/rate_limit.py:226]: i've only got a vague understanding these throttles...

if i'm a bad actor on a single IP stuffing email addresses into the login form will I get rate limited at some point?

**Review by pauldambra — APPROVED** [review]
couple of questions but 👍

**Review by webjunkie — APPROVED** [review]

File: posthog/rate_limit.py (hunk 3 above)
Comment (benjackwhite) [reply posthog/rate_limit.py:226]: Nope. Only per email after this change.

Comment (pauldambra) [reply posthog/rate_limit.py:226]: Should we have a fallback rate limit? 

It's relatively common for attackers to find sites with poor rate limits and use them to check email/common password combos to get around rate limits on the site they really want to attack...

Comment (pauldambra) [reply posthog/rate_limit.py:226]: (can be a follow up since we want to test this really fixes the issue folk are experiening sooner rather than later)

Comment (benjackwhite) [reply posthog/rate_limit.py:226]: Login is protected differently (axes). This only applies to email validation / password reset requests.

Comment (pauldambra) [reply posthog/rate_limit.py:226]: 👍 someone could still stuff emails to find valid accounts but relatively low risk :)

**Pull request merged**